### PR TITLE
[Interop][SwiftToCxx] Initial support for getting associated values from enum

### PR DESCRIPTION
**Pull request by WANGJIEKE**
This pull request adds initial support for getting associated values in struct/enum/primitive types from enums.

This pull request also rewrites some previous enum test cases because they use types from standard library or reference type, which cannot be test easily.

### CODE REVIEW

**Comment by hyp** [comment]
@swift-ci please test

**Comment by WANGJIEKE** [comment]
Solve the merge conflict.

@hyp Could you please review this PR? Thanks.

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 1]
```diff
@@ -436,6 +443,54 @@ class DeclAndTypePrinter::Implementation
...
+          os << "    ";
+          clangFuncPrinter.printClangFunctionReturnType(
+              firstType, optKind, firstTypeDecl->getModuleContext(),
+              owningPrinter.outputLang);
+          os << " result;\n";
+          os << "    "
+                "memcpy(&result, payloadFromDestruction, sizeof(result));\n";
+          os << "    return result;\n";
+        } else {
+          os << "    return " << cxx_synthesis::getCxxImplNamespaceName();
```
Comment (hyp) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:483]: This can be in another modules namespace.

**Review by hyp — COMMENTED** [review]
I’ve requested a couple of changes , please take a look

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 2]
```diff
@@ -436,6 +443,54 @@ class DeclAndTypePrinter::Implementation
...
+            getObjectTypeAndOptionality(firstTypeDecl, firstType);
+
+        // FIXME: (tongjie) may have to forward declare return type
+        os << "  inline ";
+        clangFuncPrinter.printClangFunctionReturnType(
+            firstType, optKind, firstTypeDecl->getModuleContext(),
+            owningPrinter.outputLang);
+        os << " get" << name << "() const {\n";
+        os << "    if (!is" << name << "()) abort();\n";
+        os << "    auto thisCopy = *this;\n";
```
Comment (hyp) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:469]: While this copy is correct, it incurs an extra destroy operation on the value, as you're using a C++ class value on stack, whose destructor gets called and incorrectly tries to destroy the enum after `_destructiveProjectEnumData`. We should avoid that. One idea is to use placement `new` on a buffer you allocate on stack that can store the copy C++ type. This way we can avoid calling the destructor there.

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please tes

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Review by hyp — APPROVED** [review]
LGTM

**Pull request merged**